### ISSUE TITLE
Write up technical implementation document for LPA email spike

### ISSUE THREAD

**Issue by DilwoarH** [digital-land/submit#729]
## What

Technical implementation plan for the daily LPA email actions feature

## Why

So we can present to the relevant stakeholders and get buy-in

**Comment by DilwoarH** [comment]
# Implementation Plan: Email Status Update for ODP Programme

## tldr;

This document explains how the solution will be implemented.

The following tasks need to be done:
1. Create a Google Console API project with access to Google Sheets and Google Drive APIs (~1-2 hours)
2. Create a new API endpoint for the dataset overview page to get back dataset status information (~1 day)
3. Use the information from step 2 to generate email content for each LPA (~1 day)
4. Create a way to retrieve email from a specific Google and send them their respective LPA email (~0.5 days)
5. Use GOVUK Notify to send these emails (~0.5 days)

**Spikes Reference**
- [#675](https://github.com/digital-land/submit/issues/675)  
- [#672](https://github.com/digital-land/submit/issues/672)  
- [#673](https://github.com/digital-land/submit/issues/673)  

## Background  
**Product Overview:** The ODP (Open Data Programme) Email Status Update aims to streamline communication with stakeholders by automating status updates for datasets. This involves leveraging Google Sheets for data management, Node.js for data retrieval and email dispatch, and GOV.UK Notify for secure email delivery.

---

## Implementation Details  

### 1. Requirements Analysis  
- **Data Management:** Use Google Sheets to maintain organisation details, email contacts, dataset statuses, and any related issue notes.  
- **Data Retrieval:** A Node.js application will query the Google Sheets API to extract the required information.  
- **Email Composition:** Generate personalised emails for each organisation based on their dataset statuses and reported issues.  
- **Email Dispatch:** Send emails using the GOV.UK Notify API for efficient and secure delivery.

---

### 2. System Design  

#### 2.1 Google Sheets Setup  
- **Structure:** Design a Google Sheet with the following columns:
  - Organisation ID  
  - Contact Name  
  - Contact Email

- **Access Control:** Share the sheet with a service account email for secure API access.  

#### 2.2 Node.js Application  
**Key Modules:**  
1. **Data Retrieval:** Fetch information using the Google Sheets API.  
2. **Email Generation:** Construct email content dynamically using retrieved data.  
3. **Email Sending:** Use the GOV.UK Notify API to dispatch emails.  

---

### 3. Implementation Steps  

#### 3.1 Google Console Configuration  
1. **Enable API Access:**  
   - In Google Cloud Console, enable the Google Sheets API for your project.  

2. **Service Account Setup:**  
   - Create a service account with appropriate permissions.  
   - Download the JSON key file for authentication.  
   - Share the Google Sheet with the service account’s email.  

#### 3.2 Node.js Application Development  

1. **Dependency Installation:**  
   - Install Google APIs client library:  
     ```bash
     npm install googleapis
     ```  
   - Install GOV.UK Notify client library:  
     ```bash
     npm install notifications-node-client
     ```  

2. **Data Retrieval Module:**  
   - Authenticate using the service account key.  
   - Fetch and process data from the Google Sheet.  

   **Example Code:**  
```javascript
const { google } = require('googleapis');
const path = require('path');

// Path to your service account key file
const SERVICE_ACCOUNT_FILE = path.join(__dirname, 'service-account-key.json');

// The ID of your Google Sheet
const SHEET_ID = '1Qe5_EOQNJlMNVHcyLhLTKvnDppDxypdYF1b7D0zz3ME'; // Replace with your Google Sheet ID

// The range of cells to read (e.g., "Sheet1!A1:E10")
const RANGE = 'Sheet1!A1:B100';

// Function to authenticate and fetch data
async function readGoogleSheet() {
  try {
    // Authenticate with the service account
    const auth = new google.auth.GoogleAuth({
      keyFile: SERVICE_ACCOUNT_FILE,
      scopes: ['https://www.googleapis.com/auth/spreadsheets.readonly'],
    });

    // Create the Sheets API client
    const sheets = google.sheets({ version: 'v4', auth });

    // Fetch the data
    const response = await sheets.spreadsheets.values.get({
      spreadsheetId: SHEET_ID,
      range: RANGE,
    });

    // Log the data to the console
    const rows = response.data.values;
    if (rows.length) {
      console.log('Data from the sheet:');
      rows.forEach((row) => {
        console.log(row);
      });
    } else {
      console.log('No data found in the specified range.');
    }
  } catch (error) {
    console.error('Error reading Google Sheet:', error.message);
  }
}

// Call the function
readGoogleSheet();
```

3. **Email Generation Module:**
  - Define a dynamic email template with placeholders.
  - Populate the template with data relevant to each organisation.

4. **Email Sending Module:**
  - Configure the GOV.UK Notify client with an API key.
  - Dispatch emails with personalised content.

  **Example Code:**

```js
const NotifyClient = require('notifications-node-client').NotifyClient;

const notifyClient = new NotifyClient("YOUR_NOTIFY_API_KEY");
const templateId = "YOUR_TEMPLATE_ID";

const emails = ["contact1@example.com", "contact2@example.com"];

emails.forEach(email => {
    notifyClient.sendEmail(templateId, email, {
        personalisation: { key1: "value1", key2: "value2" },
        reference: "unique-reference"
    }).then(response => console.log(response))
      .catch(error => console.error(error));
});
```

#### 3.3 **Testing**
1. Unit Testing:
  - Test modules for data retrieval, email composition, and dispatch functionality independently.
2. Integration Testing:
  - Validate the end-to-end process from data retrieval to email dispatch.
3. User Acceptance Testing:
  - Confirm email content accuracy and adherence to stakeholder requirements.

#### 3.4 **Deployment**
  - Hosting: Code should be hosted in the same [submit](https://github.com/digital-land/submit) repo.
  - Scheduling: Use a task scheduler like cron to automate periodic email dispatches.

--- 
 
### 4. **Maintenance and Monitoring**
  - Error Handling: Implement robust error logging for data retrieval and email sending.
  - Monitoring:
    - Review application logs and the GOV.UK Notify dashboard to ensure email delivery success.
    - Updates: Periodically review and update dependencies, and verify the Google Sheet structure remains aligned with system requirements.

**Comment by DilwoarH** [comment]
@coderabbitai - create a sequence diagram based on the above comment

**Comment by MJFaucherFolie** [comment]
Looks good on my end! Love the CodeRabbit elements. The next steps would be to play this back to Infra to agree approach moving forward?

**Comment by MJFaucherFolie** [comment]
Scheduled a call with Infra on Tuesday 17th Dec to play back findings 

FYI: @DilwoarH

**Comment by MJFaucherFolie** [comment]
@DilwoarH played back his technical implementation document to the Infra team on Tuesday 17th December.

---

## Write up
- The simplest approach is to set up a Lamda function that Airflow triggers
- As this is not just a provider's task, we should involve members of the Infrastructure team. They will attend our stand-up and provide updates.
- There is a CRM tool which has been developed, so integrating with this (in the near future) is the best approach
- Must ensure that the messages being sent to LPAs matches the specifications, as we are aware that the dashboard shows issues against items that are optional (I.E - Blank name, blank description fields). 

## Tasks
- Owen - To playback to PSD the solution we are proposing and get his input.
- MJ - Investigate if JIRA is the place to hold CRM data.
- MJ - Discuss with @neilfwar if we could integrate with the CRM tool and next steps to achieving this.
- Gordon - Request that Samriti attends our stand-up when the work starts.

**Comment by MJFaucherFolie** [comment]
## Jira
- Raised this in the Jira Help Slack channel
- Confirmed that because we have access to the JIRA API, it is feasible for us to use JIRA as a CRM

## CRM tool
- Had a brief chat with Neil regarding the CRM tool he was involved with. 
- MS Power Apps allows [API connections](https://www.google.co.uk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwi7sYn0-q6KAxUgTqQEHYLINYQQma4LegQIPhAC&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DAac8LTo9eIA%26t%3D377&usg=AOvVaw06AH45GwuGxrVC00P7Rxkb&opi=89978449), but it was advised to inform the Product Manager of the CRM tool, so it can be worked into their SoW

**Comment by MJFaucherFolie** [comment]
GOV.UK Notify's analytics is quite lacklustre and will hamper our ability to assess the true impact of our email messaging. @neilfwar has conducted analysis of a email tool called Send In Blue also known as Brevo, The tool has the following features:

Open rate
- The percentage of recipients who opened your email
Click-through rate (CTR)
- The percentage of recipients who clicked a link in your email after opening it
Unsubscribe rate
- The percentage of recipients who opt-out of your email list after receiving an email
Conversion rate
- The percentage of recipients who clicked on a link in your email and took the intended action, such as making a purchase or filling out a form
Bounce rate
- The percentage of emails sent that were not delivered to the intended recipient

**Comment by Ben-Hodgkiss** [comment]
@MJFaucherFolie @CharliePatterson 

Earlier today, @eveleighoj and I met to discuss this. We've created a sub-task (https://github.com/digital-land/technical-documentation/issues/203) which contains the work that Infrastructure will need to do. There are a couple of outstanding questions that we have - these can be found in the blue sticky notes of our ticket planning Mural (sent to you both directly). 

These are:
- We need to ensure that however the emails are being managed there is a permanent plan for where they come from and are kept up-to-date.
- Will the data required for the email (e.g. number of errors, which datasets etc) be provided by a submit endpoint or somewhere else?
- Do we need to do any filtering to only send email to those LPAs who need it?

**Comment by MJFaucherFolie** [comment]
> [@MJFaucherFolie](https://github.com/MJFaucherFolie) [@CharliePatterson](https://github.com/CharliePatterson)
> 
> Earlier today, [@eveleighoj](https://github.com/eveleighoj) and I met to discuss this. We've created a sub-task ([digital-land/technical-documentation#203](https://github.com/digital-land/technical-documentation/issues/203)) which contains the work that Infrastructure will need to do. There are a couple of outstanding questions that we have - these can be found in the blue sticky notes of our ticket planning Mural (sent to you both directly).
> 
> These are:
> 
> * We need to ensure that however the emails are being managed there is a permanent plan for where they come from and are kept up-to-date.
> * Will the data required for the email (e.g. number of errors, which datasets etc) be provided by a submit endpoint or somewhere else?
> * Do we need to do any filtering to only send email to those LPAs who need it?

@CharliePatterson and I have reviewed your questions and [prepared a doc](https://docs.google.com/document/d/1YKe_gZB8Mtqi3FppD4Tt25upfsXr3p3PeZLOEgUO13U/edit?usp=sharing) detailing our responses :)
